hs_core/views/resource_rest_api.py

```diff
@@ -893,6 +897,130 @@ def _validate_metadata(metadata_list):
+            # existing_path_in_resource is a list of folder names
+            # append them into a path
+            for folder in existing_path_in_resource:
+                eventual_relative_path += folder + '/'
+    except Exception as ex:
+        logger.info(f"Existing path in resource not found: {str(ex)}")
+
+    # handle the case that a folder was uploaded instead of a single file
+    # use the metadata.relativePath to rebuild the folder structure
+    path_within_uploaded_folder = metadata.get('relativePath', '')
+    # path_within_resource_contents will include the name of the file, so we need to remove it
+    path_within_uploaded_folder = os.path.dirname(path_within_uploaded_folder)
+    if path_within_uploaded_folder:
+        eventual_relative_path += path_within_uploaded_folder
+    path = f'{hs_res_id}/data/contents/{eventual_relative_path}'
+    return path
+
+
+class CustomTusFile(TusFile):
+    # extend TusFile to allow it to write to s3 storage instead of local disk
+    # https://github.com/alican/django-tus/blob/2aac2e7c0e6bac79a1cb07721947a48d9cc40ec8/django_tus/tusfile.py#L52
+
+    def __init__(self, resource_id):
+        self.storage = S3Storage()
+        self.resource_id = resource_id
+        self.filename = cache.get("tus-uploads/{}/filename".format(resource_id))
+        self.file_size = int(cache.get("tus-uploads/{}/file_size".format(resource_id)))
+        self.metadata = cache.get("tus-uploads/{}/metadata".format(resource_id))
+        self.offset = cache.get("tus-uploads/{}/offset".format(resource_id))
+        bucket, _ = bucket_and_name(self.metadata.get("hs_res_id"))
+        self.bucket = bucket
+        self.path = get_path(self.metadata)
+
+    def get_storage(self):
+        return self.storage
+
+    @staticmethod
+    def create_initial_file(metadata, file_size):
+        resource_id = str(uuid.uuid4())
+        cache.add("tus-uploads/{}/filename".format(resource_id),
+                  "{}".format(metadata.get("filename")), settings.TUS_TIMEOUT)
+        cache.add("tus-uploads/{}/file_size".format(resource_id), file_size, settings.TUS_TIMEOUT)
+        cache.add("tus-uploads/{}/offset".format(resource_id), 0, settings.TUS_TIMEOUT)
+        cache.add("tus-uploads/{}/metadata".format(resource_id), metadata, settings.TUS_TIMEOUT)
+
+        tus_file = CustomTusFile(resource_id)
+        tus_file.write_init_file()
+        return tus_file
+
+    def get_path_and_id(self):
+        if not self.path:
+            self.path = get_path(self.metadata)
+        return self.path + self.resource_id
+
+    def get_path_and_name(self):
+        if not self.path:
+            self.path = get_path(self.metadata)
+        return self.path + self.filename
+
+    @staticmethod
+    def check_existing_file(path):
+        # TODO: #5685 this seems to return False even when the file exists...
+        # need to add bucket?
+        return S3Storage().exists(path)
```

Comment:
@sblack-usu is there a different way that I should be checking to see if an object exists?

Reply:
this path should not include the bucket